client.go

```diff
@@ -520,6 +526,269 @@ func (c *Client) ListImages(ctx context.Context, filters ...string) ([]Image, er
 	return images, nil
 }
 
+// Checkpoint creates a checkpoint of the container
+func (c *Client) Checkpoint(ctx context.Context, id, ref string, opts ...CheckpointOpts) error {
+	index := &ocispec.Index{
+		Versioned: ver.Versioned{
+			SchemaVersion: 2,
+		},
+		Annotations: make(map[string]string),
+	}
+	copts := &options.CheckpointOptions{
+		Exit:                false,
+		OpenTcp:             false,
+		ExternalUnixSockets: false,
+		Terminal:            false,
+		FileLocks:           true,
+		EmptyNamespaces:     nil,
+	}
+	container, err := c.LoadContainer(ctx, id)
+	if err != nil {
+		return err
+	}
+	img, err := container.Image(ctx)
+	if err != nil {
+		return err
+	}
+	index.Annotations["image.name"] = img.Name()
+
+	ctx, done, err := c.WithLease(ctx)
+	if err != nil {
+		return err
+	}
+	defer done(ctx)
+
+	// pause task to checkpoint
+	if err := func(ctx context.Context) error {
+		task, err := container.Task(ctx, nil)
+		if err != nil {
+			if errdefs.IsNotFound(err) {
+				return nil
+			}
+			return err
+		}
+		if err := task.Pause(ctx); err != nil {
+			return err
+		}
+		defer task.Resume(ctx)
+
+		info, err := container.Info(ctx)
+		if err != nil {
+			return err
+		}
+
+		// add runtime info to index
+		index.Annotations["runtime.name"] = info.Runtime.Name
+		if info.Runtime.Options != nil {
+			data, err := info.Runtime.Options.Marshal()
+			if err != nil {
+				return err
+			}
+			r := bytes.NewReader(data)
+			desc, err := writeContent(ctx, c.ContentStore(), images.MediaTypeContainerd1CheckpointRuntimeOptions, info.ID+"-runtime-options", r)
+			if err != nil {
+				return err
+			}
+			desc.Platform = &ocispec.Platform{
+				OS:           runtime.GOOS,
+				Architecture: runtime.GOARCH,
+			}
+			index.Manifests = append(index.Manifests, desc)
+		}
+
+		// process remaining opts
+		for _, o := range opts {
+			if err := o(ctx, c, &info, index, copts); err != nil {
+				return err
+			}
+		}
+		return nil
+	}(ctx); err != nil {
+		return err
+	}
+
+	desc, err := writeIndex(ctx, index, c, container.ID()+"index")
+	if err != nil {
+		return err
+	}
+	i := images.Image{
+		Name:   ref,
+		Target: desc,
+	}
+	if _, err := c.ImageService().Create(ctx, i); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Restore restores a container from a checkpoint
+func (c *Client) Restore(ctx context.Context, id, ref string, live bool) error {
```

Comment:
Need restore Opts, not `live bool`  There should be handlers for each media type on how to restore it or things to omit

Reply:
👍 that's how i had it before but with a single opt.  thx